agent/src/runtime_connectors/podman_cli.rs

```diff
@@ -53,6 +55,73 @@ impl From<PodmanContainerInfo> for ExecutionState {
     }
 }
 
+struct TimedPodmanPsResult(Mutex<Option<(time::Instant, Arc<PodmanPsResult>)>>);
+
+impl TimedPodmanPsResult {
+    async fn get(&self) -> Arc<PodmanPsResult> {
+        let mut guard = self.lock().await;
+
+        if let Some(value) = &mut *guard {
+            if value.0.elapsed().as_millis() > PODMAN_PS_CACHE_MAX_AGE {
+                *value = Self::new_inner().await;
+            }
+            value.1.clone()
+        } else {
+            let ps_result = Self::new_inner().await;
+            let result = ps_result.1.clone();
+            *guard = Some(ps_result);
+            result
+        }
+    }
+
+    async fn new_inner() -> (time::Instant, Arc<PodmanPsResult>) {
+        (
+            time::Instant::now(),
+            Arc::new(PodmanCli::list_states_internal().await.into()),
+        )
+    }
+}
+
+impl Deref for TimedPodmanPsResult {
+    type Target = Mutex<Option<(time::Instant, Arc<PodmanPsResult>)>>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
```

Comment:
Note before reading: This is just a question. We don't need to change anything.
We could have also used a type definition, not that it matters. Was there a particular reason to do it via a function?

Reply:
I have used to Newtype pattern to be able to write an `impl` block for `TimedPodmanPsResult`.

As alternative I could have defined a trait and implement this one for `Mutex<Option<(time::Instant, Arc<PodmanPsResult>)>>`.
I did not want to do this as:

- methods have to be defined twice (once in `trait`, once in `impl`)
- helper methods cannot be defined in trait `impl` blocks

I thinks this Deref implementation does not cause any problems:

- it does not disturb the reading of the implementation of the actual logic.
- if the compiler optimisation is not set to 0, this code should simply disappear.

Reply:
Alright